src/NameGenerator.elm

```diff
@@ -0,0 +1,116 @@
+module NameGenerator exposing
+    ( Definition, Lexicon
+    , generator
+    , capitalize
+    , fromString
+    )
+
+
+import Dict exposing (Dict)
+import Random.Pcg as Random exposing (Generator)
+import String
+import Regex as R
+
+
+
+-- types
+
+type Fragment
+    = Constant String
+    | Key String
+
+
+type alias Definition =
+    List Fragment
+
+
+type alias Lexicon =
+    Dict String (List Definition)
+
+
+
+-- Generate a name given a lexicon and a key of that lexicon
+
+generator : Lexicon -> String -> Generator String
+generator lexicon key =
+
+    case Dict.get key lexicon of
+
+        Nothing ->
+            -- either the key is plain invalid, or it is stuck in a loop
+            Random.constant <| "--[" ++ key ++ "]--"
+
+        Just definitions ->
+            let
+                reducedLexicon =
+                    Dict.remove key lexicon
```

Comment:
Is this the best way to kill loops?
Would passing the key stack be feasible *and* better?